WebDriverAgentLib/Routing/FBResponseJSONPayload.m

```diff
@@ -20,17 +20,77 @@ @interface FBResponseJSONPayload ()
 
 @implementation FBResponseJSONPayload
 
+
+/*
+ * Convenience method to do the check and validation in one.
+ */
++ (NSString*) makeValidUTF8:(NSString*) stringToCheck
+{
+  if (![FBResponseJSONPayload isValidUTF8 :stringToCheck])
+  {
+    return [FBResponseJSONPayload removeInvalidCharsFromString:stringToCheck];
+  }
+  else
+  {
+    return stringToCheck;
+  }
+}
+
+/*
+ * Returns true if the string can be converted to UTF8
+ */
++ (int) isValidUTF8:(NSString*) stringToCheck
+{
+  return ([stringToCheck UTF8String] != nil);
+}
+
+/*
+ * Removes invalid UTF8 chars from the NSString
+ * This method is slow, so only run it on strings that fail the isValidUTF8 check.
+ */
++ (NSString*) removeInvalidCharsFromString:(NSString*) stringToCheck
+{
+  NSMutableString* fixedUp = [NSMutableString stringWithString:@""];
+  for (NSUInteger i = 0; i < [stringToCheck length]; i++)
+  {
+    @autoreleasepool
+    {
+      unichar character = [stringToCheck characterAtIndex:i];
+      NSString* charString = [[NSString alloc] initWithCharacters:&character length:1];
+      if ([charString UTF8String] == nil)
+      {
+        [FBLogger logFmt:@"Invalid UTF-8 sequence encountered at position %lu. Code: %hu (%X). Replacing with \ufffd", (unsigned long) i, character, character];
+        [fixedUp appendString:@"\ufffd"];
+      }
+      else
+      {
+        [fixedUp appendString:charString];
+      }
+    }
+  }
+  [FBLogger logFmt:@"Given JSONPayload was NOT valid utf-8. Orig length %lu, fixed length %lu", (unsigned long)[stringToCheck length], (unsigned long)[fixedUp length]];
+  return fixedUp;
+}
+
 - (instancetype)initWithDictionary:(NSDictionary *)dictionary
                     httpStatusCode:(HTTPStatusCode)httpStatusCode
 {
   NSParameterAssert(dictionary);
   if (!dictionary) {
     return nil;
   }
-
+  NSMutableDictionary* newDictionary = [NSMutableDictionary dictionary];
```

Comment:
I would rather make this check local for the getSource endpoint only. Changing it for all endpoints might break backwards compatibility

Reply:
also from the compatibility perspective it makes sense to put this behavior under a setting and disable it by default.

Reply:
I agree on you to check it only on getSource endpoint, and sure, I can add another settings. Would you like `replaceUnsupportedUnicodeCharcters` for the name of a setting?